ci-operator/step-registry/cucushift/installer/rehearse/vsphere/upi/provision/cucushift-installer-rehearse-vsphere-upi-provision-chain.yaml

```diff
@@ -12,8 +12,10 @@ chain:
   - ref: upi-conf-vsphere-dns
   - ref: upi-conf-vsphere-ova
   - ref: upi-conf-vsphere-ova-vcm
+  - ref: ipi-conf-vsphere-folder
   - ref: upi-install-vsphere
   - ref: enable-qe-catalogsource
   - chain: cucushift-installer-check
+  - ref: cucushift-installer-check-vsphere-folder
```

Comment:
Do you know once we enable customer folder for UPI, how UPI installation launch VMs in the folder? As far as I know, UPI is using some script to launch vms, I did not find out where the UPI script reference the custom folder.

Reply:
And from https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/57129/rehearse-57129-periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-upi-encrypt-f28/1838870980630417408/artifacts/vsphere-upi-encrypt-f28/cucushift-installer-check-vsphere-folder/build-log.txt, I did not see the check folder step is executed.

Reply:
I think it inherited form install config and set it here like https://github.com/openshift/installer/blob/30ba9c57d53b8519d158d748bb4c87a82ff8fadc/upi/vsphere/powercli/upi-functions.ps1#L312

Reply:
the install-config.yaml will be used to generated as ignition, then the ignition will be injected to vm as metadata when it is being created, so I do not think the line is getting the customized folder, I think that line is just to get datacenter folder. Let us wait for your test result to see what happen.

Reply:
upi cluster created as expected in job https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/57129/rehearse-57129-periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-upi-encrypt-f28/1839518910836117504

![image](https://github.com/user-attachments/assets/3b6d9c22-5f38-4c60-ab00-4caaf32ad5c5)

Reply:
From https://github.com/openshift/installer/blob/30ba9c57d53b8519d158d748bb4c87a82ff8fadc/upi/vsphere/powercli/upi.ps1#L178-L188, seems that prowshell always creates folder `$($fd.datacenter)/$($fd.cluster)/vm/${cluster_name}` if it does not exist, but not use any custom folder defined in install-config.

And I remembered that you filed bug for this before, and searched this one https://issues.redhat.com/browse/OCPBUGS-35489, from dev comments, looks like there is no plan to support custom folder in upi poweshell. Is that right? Or I missed something?

Reply:
Agree with jinyun.

Per the upi install [log](https://gcsweb-qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/qe-private-deck/pr-logs/pull/openshift_release/57129/rehearse-57129-periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-upi-encrypt-f28/1839518910836117504/artifacts/vsphere-upi-encrypt-f28/upi-install-vsphere/build-log.txt), 
```
Getting viserver for vcenter.devqe.ibmc.devcluster.openshift.com
Checking for folder in failure domain DEVQEdatacenter/DEVQEcluster
Checking for resource pool in failure domain DEVQEdatacenter/DEVQEcluster
```
That means the upi script expected folder already exists, so did not create a new one, but that does not mean the folder setting in install-config.yaml is respected. In the [PR](https://github.com/openshift/release/pull/57129/files#diff-599e7e43342aa05af9c956fd20f189b177452ab2b8d26258f2c32a07ce8bd384R32), in check script, the expecting folder path is `"/$vsphere_datacenter/vm/$vsphere_datacenter/${NAMESPACE}-${UNIQUE_HASH}"`, there is no `vm` folder path is displayed in the screen shot, is that true ?

Reply:
Got it, name folder with ${NAMESPACE}-${UNIQUE_HASH} maybe the reason, I'll run a destruct job, where folder name is not default.

what wired is cluster indeed created in /$vsphere_datacenter/vm/$vsphere_datacenter

![image](https://github.com/user-attachments/assets/d7e1f3b6-fb54-4297-9590-8e4ef34e7d09)

Reply:
Hmm, `Get-Folder -Location $folder` seems retries all folders matched with given name in the specified location, see https://vdc-repo.vmware.com/vmwb-repository/dcr-public/0a01c8b1-4515-46e5-ade9-d457877d167e/f63b3243-b9c0-4199-bfc9-8a6b3b11151a/doc/Get-Folder.html
So pwsh does not create new folder as it already exits, your post-check can also pass.

I guess post-check will be failed if you don't use `cluster name` as your custom folder name.

Reply:
> what wired is cluster indeed created in /$vsphere_datacenter/vm/$vsphere_datacenter
Per https://github.com/openshift/installer/blob/30ba9c57d53b8519d158d748bb4c87a82ff8fadc/upi/vsphere/powercli/upi-functions.ps1#L306-L312, I guess the key line is:
```
   $folder = Get-Folder -Server $node.server -Name $clustername -Location $node.datacente
```
Maybe the command has some "search" functionality, if we pre-created a folder like `/$vsphere_datacenter/vm/$vsphere_datacenter/$cluster-name`, then the command always can located it by "searching" cluster-name, but do not need the accurate full folder path, that is why your testing got passed. But actually it is not from install-config.yaml setting. For example, if we pre-created a folder named "mytesting", and set it to install-config.yaml, your testing would fail.

Reply:
> I guess post-check will be failed if you don't use cluster name as your custom folder name.

Just saw the comments, totally the same understanding from me.

Reply:
yes, see job https://qe-private-deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gs/qe-private-deck/pr-logs/pull/openshift_release/57129/rehearse-57129-periodic-ci-openshift-openshift-tests-private-release-4.17-amd64-nightly-vsphere-upi-encrypt-f28-destructive/1839560601542594560

It does not use folder qe-upi-folder

![image](https://github.com/user-attachments/assets/06917237-e39d-4d27-ae11-79766d80b72f)

Reply:
So what to do about this PR, shall I remove UPI part?

Reply:
per @jinyunma mentioned bug - OCPBUGS-35489 in https://github.com/openshift/release/pull/57129#discussion_r1778073723, upi + custom folder is never supported. So I think this case is an invalid case, should be disabled, of course, if the custom folder is just one of multi test point in the case, then we need remove only the custom folder test point from the case.

Reply:
Got it, I'll disable the case and close this PR, thanks for the help

Reply:
> check that folder in configmap cloud-provider-config on running cluster should be correct with absolute path.

So for folder testing, even in IPI install, there are two test points:
1. use a custom folder path with sub-folder, e.g: a/b, but not a
2. in check step, check the folder is correct in cloud-provider-config configmap.

For 1, per https://github.com/openshift/release/pull/57129/files#diff-a4e33b1c984da8649ad7d33b8c1d24e1b23a9d734ad7b23248b34c2af845c565R30, the current sub-folder is `$vsphere_datacenter/${NAMESPACE}-${UNIQUE_HASH}`, I suggest to use some special folder name, e.g: `qe-testing/ci-${NAMESPACE}-${UNIQUE_HASH}-cluster` to avoid vsphere datacenter name and cluster name.

For 2, sounds like we missed the test step in the check folder step, we need to add it.

Reply:
1, I can change subfolder to ci-${UNIQUE_HASH}-cluster, folder with the same name of datacenter is a test case, see https://polarion.engineering.redhat.com/polarion/#/project/OSE/workitem?id=OCP-30630 
2, will add cloud-provider-config check point to IPI